src/data/equiv/algebra.lean

```diff
@@ -356,3 +356,124 @@ instance symm.is_ring_hom {e : α ≃r β} : is_ring_hom e.to_equiv.symm := hom
+λ g h, mul_equiv.trans h g.
+This means that multiplication agrees with composition, (g*h)(x) = g (h x) .
+-/
+instance aut_group (γ : Type*) [has_mul γ] : group (aut γ) :=
+{ mul := λ g h, mul_equiv.trans h g,
+  one := mul_equiv.refl γ,
+  inv := mul_equiv.symm,
+  mul_assoc := λ _ _ _, by { ext, refl },
+  one_mul := λ _, by { ext, refl },
+  mul_one := λ _, by { ext, refl },
+  mul_left_inv := λ _, by { ext, apply equiv.left_inv } }
+
+end monoid
+
+namespace group
+
+-- A group homomorphism is a monoid homomorphism between groups.
+def aut (γ : Type*) [group γ] := mul_equiv γ γ
+
+instance aut_group (γ : Type*) [group γ] : group (aut γ) := monoid.aut_group γ
+
+end group
+
+
+@[extensionality] lemma add_equiv.ext {α β : Type*} [has_add α] [has_add β]
+  {f g : add_equiv α β} (h : f.to_fun = g.to_fun) : f = g :=
+begin
+  have h₁ := @equiv.eq_of_to_fun_eq _ _ f.to_equiv g.to_equiv h,
+  cases f, cases g, congr,
+  { exact h },
+  { exact congr_arg equiv.inv_fun h₁ }
+end
+
+namespace add_monoid
+
+def aut (α : Type*) [has_add α] := add_equiv α α
+
+/--
+The group operation on automorphisms of an additive monoid is defined by
+λ g h, add_equiv.trans h g.
+This means that multiplication agrees with composition, (g*h)(x) = g (h x) .
+-/
+instance aut_group (α : Type*) [has_add α] : group (aut α) :=
+{ mul := λ g h, add_equiv.trans h g,
+  one := add_equiv.refl α,
+  inv := add_equiv.symm,
+  mul_assoc := λ _ _ _, by { ext, refl },
+  one_mul := λ _, by { ext, refl },
+  mul_one := λ _, by { ext, refl },
+  mul_left_inv := λ _, by { ext, apply equiv.left_inv } }
+
+end add_monoid
+
+namespace add_group
+
+-- An additive group homomorphism is an additive monoid homomorphism between groups.
+def aut (γ : Type*) [add_group γ] := add_equiv γ γ
+
+instance aut_group (γ : Type*) [add_group γ] : group (aut γ) := add_monoid.aut_group γ
+
+end add_group
+
+
+@[extensionality] lemma ring_equiv.ext {R S : Type*} [ring R] [ring S]
+  {f g : ring_equiv R S} (h : f.to_fun = g.to_fun) : f = g :=
+begin
+  have h₁ := @equiv.eq_of_to_fun_eq _ _ f.to_equiv g.to_equiv h,
+  cases f, cases g, congr,
+  { exact h },
+  { exact congr_arg equiv.inv_fun h₁ }
+end
+
+namespace ring
+
+def aut (R : Type*) [ring R] := ring_equiv R R
+
+/--
+The group operation on automorphisms of a ring is defined by
+λ g h, ring_equiv.trans h g.
+This means that multiplication agrees with composition, (g*h)(x) = g (h x) .
+-/
+instance aut_group (R : Type*) [ring R] : group (aut R) :=
+{ mul := λ g h, ring_equiv.trans h g,
+  one := ring_equiv.refl R,
+  inv := ring_equiv.symm,
+  mul_assoc := λ _ _ _, by { ext, refl },
+  one_mul := λ _, by { ext, refl },
+  mul_one := λ _, by { ext, refl },
+  mul_left_inv := λ _, by { ext, apply equiv.left_inv } }
+
+/--
+Map from the automorphisms of a ring to the automorphisms of the additive
+group defined by addition in R.
+-/
+def ring_aut_to_add_monoid_aut (R : Type*) [ring R] (f : aut R) : add_monoid.aut R :=
```

Comment:
Proof that this is a group homomorphism?

Reply:
It doesn't seem like you can prove that this is a group homomorphism unless all the ```Type*```s are changed to ```Type```s.

Reply:
fixed